### ISSUE TITLE
STEP: Export to inch is creating a model in mm (Possibly fixed in OCC 7.9.0)

### ISSUE THREAD

**Issue by vernzimm** [FreeCAD/FreeCAD#18078]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

I have Step file export unit set to inches. With newest nightly build it exports the geometry data (the actual numerical information) as millimeters, but still tags the file as inch, meaning the model scales up in size. Previous nightly build I was using was 38684.

I do see there was an update from Open CASCADE 7.7.2 to 7.8.1. I did take a look at their tracker as well, but don't see any mention of an issue like this.

***Edit: I tracked the issue down to a known bug in OCCT 7.8.1 [OCCT Issue 354](https://github.com/Open-Cascade-SAS/OCCT/issues/354) that they fixed with 7.9.***

### Full version info

```shell
OS: Windows 11 build 22631
Architecture: x86_64
Version: 1.1.0dev.39240 (Git) Conda
Build type: Release
Branch: main
Hash: f9ba94c70b139ece6ba5ffba47632d91b5c9c645
Python 3.11.10, Qt 5.15.15, Coin 4.0.3, Vtk 9.3.0, OCC 7.8.1
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/Qt default
Installed mods: 
  * Behave-Dark-Colors 0.1.1
  * ProDarkThemePreferencePack 1.0.0
  * ThreadProfile 1.92.0
```

### Subproject(s) affected?

Core

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by tarman3** [comment]
Confirm with 1.0
Do not confirm with weekly-builds-39240

**Comment by vernzimm** [comment]
No, I made the model from scratch from a sketch in 39240.

I do notice though that your OCC version for 39240 is 7.7.2?

**Comment by tarman3** [comment]
I will try to download 1.0.0_appimage and check.

**Comment by tarman3** [comment]
Ok, to be clear write my test.
- Open FreeCAD
- File > New
- Open workbench Part
- Create default Cube
- Select Cube
- File > Export > Step with colors (*.step, *.stp) > Units for export of STEP = Inch
- File > Import > All settings by default

Issue **Confirmed**
```
OS: Arch Linux (KDE/plasma/wayland)
Architecture: x86_64
Version: 1.0.0.39109 (Git)
Build type: Release
Branch: makepkg
Hash: 2fcc5317fe3aee96ca73475986a577719fc78e20
Python 3.12.7, Qt 6.8.0, Coin 4.0.3, Vtk 9.3.1, OCC 7.8.1
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/
Installed mods: lattice2 1.0.0; CurvedShapes 1.0.13; MeshRemodel 1.10.34; freecad.gears 1.3.0; fasteners 0.5.31; sheetmetal 0.5.7; Silk 0.1.5; PieMenu 1.9.5 (Disabled); Curves 0.6.51
```

Issue **NOT** confirmed
```OS: Arch Linux (KDE/plasma/xcb)
Architecture: x86_64
Version: 1.0.0.39109 (Git) Conda AppImage
Build type: Release
Branch: (HEAD detached at 1.0.0)
Hash: 2fcc5317fe3aee96ca73475986a577719fc78e20
Python 3.11.9, Qt 5.15.13, Coin 4.0.3, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/Fusion
Installed mods: lattice2 1.0.0; CurvedShapes 1.0.13; MeshRemodel 1.10.34; freecad.gears 1.3.0; fasteners 0.5.31; sheetmetal 0.5.7; Silk 0.1.5; PieMenu 1.9.5 (Disabled); Curves 0.6.51
```


Issue **NOT** confirmed
```
OS: Arch Linux (KDE/plasma/xcb)
Architecture: x86_64
Version: 1.1.0dev.39240 (Git) Conda AppImage
Build type: Release
Branch: main
Hash: f9ba94c70b139ece6ba5ffba47632d91b5c9c645
Python 3.11.9, Qt 5.15.13, Coin 4.0.3, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/Fusion
Installed mods: lattice2 1.0.0; CurvedShapes 1.0.13; MeshRemodel 1.10.34; freecad.gears 1.3.0; fasteners 0.5.31; sheetmetal 0.5.7; Silk 0.1.5; PieMenu 1.9.5 (Disabled); Curves 0.6.51
```

**Comment by tarman3** [comment]
I look inside step files.

It seems that the versions _1.0.0.39109 (Git) Conda AppImage_ and  _1.0.0.39109 (Git) Conda AppImage_ ignore the Inch preference.
So it seems to be working fine.
But inside the files the parameters are indicated in mm, not inches.

**Comment by vernzimm** [comment]
Note Open CASCADE (OCC) versions again. Issue confirmed with OCC 7.8.1, issue not confirmed with OCC 7.7.2.

**Comment by tarman3** [comment]
Also confirmed in Linuxmint 21.3 with **OCC 7.8.1**
```
OS: Linux Mint 21.3 (X-Cinnamon/cinnamon/xcb)
Architecture: x86_64
Version: 1.1.0dev.39256 (Git)
Build type: Release
Branch: main
Hash: 6f23f01e509348a6755ad3c465a3d7ffd758ee03
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 9.1.0, OCC 7.8.1
Locale: C/Default (C)
Stylesheet/Theme/QtStyle: unset/unset/Qt default
```

**Comment by cweickhmann** [comment]
I can confirm a similar issue with the current FreeCAD AppImage and OCC 7.7.2:

1. Create a Body and a Cylinder object with dimensions: R = 100 mm and H = 50 mm.
2.A.1. Export as STEP selecting Units to be Millimetres
2.A.2. Open STEP file in Gmsh and observe a radius of 100 and a height of 50. That's okay. But:
2.B.1. Export as STEP selecting Units to be Metres
2.B.2. Open STEP file in Gmsh and observe a radius of still 100 and a height of 50. That should be 100e-3 and 50e-3, respectively.

FreeCAD Version Details:
```
OS: Ubuntu 24.04.1 LTS (KDE/plasma/xcb)
Architecture: x86_64
Version: 1.0.0.39109 (Git) Conda AppImage
Build type: Release
Branch: (HEAD detached at 1.0.0)
Hash: 2fcc5317fe3aee96ca73475986a577719fc78e20
Python 3.11.9, Qt 5.15.13, Coin 4.0.3, Vtk 9.2.6, OCC 7.7.2
Locale: German/Germany (de_DE)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/Fusion
```

OCC Version according to library listing:
```
Open CASCADE Technology
Version: 7.7.2
```

Host OS/arch:
```
Distributor ID: Ubuntu
Description:    Ubuntu 24.04.1 LTS
Release:        24.04
Codename:       noble
Kernel: Linux 6.8.0-49-generic
Arch: x86_64
```

**Comment by vernzimm** [comment]
@cweickhmann interesting. I personally would never have need to export in meters vs mm :)

I checked this with most recent nightly and the issue is still there, but I found the culprit on OCCT bugtracker...

[https://github.com/Open-Cascade-SAS/OCCT/issues/354](url)

It looks like they are shooting for 02/17/25 for release of 7.9.0. Can this be added as a goal for FreeCAD to implement new OCCT version soon when it is released?

**Comment by cweickhmann** [comment]
Well, the use case is pretty straight forward: when you use FreeCAD to generate geometries for simulations, you want to know your units. :-) I don't care too much if it's mm or m, but it should be what's written on the package.

**Comment by vernzimm** [comment]
OCCT 7.9.0 has been released, likely fixing this issue. Please upgrade when possible/feasible.

https://github.com/Open-Cascade-SAS/OCCT/releases/tag/V7_9_0

**Comment by vernzimm** [comment]
They are actually close to releasing 8.0.0... but FreeCAD most recent build is still using bugged 7.8.1? Anyone? I have been using build  38684 since Sept 2024 because I need to export .step files in the correct proportions... Anyone?

[OCCT Releases](https://github.com/Open-Cascade-SAS/OCCT/releases)

**Comment by vernzimm** [comment]
I'm not sure I follow, but maybe I'm misunderstanding. It seems to be complete and released by OCCT? I do understand it's not a drop-in upgrade, there are some changes that need to be made on FreeCAD's side...

**Comment by oursland** [comment]
FreeCAD uses a lot of package dependencies, each of those must be made compatible with OCCT 7.9.0 before FreeCAD can use OCCT 7.9.0.  That has not yet been completed.

**Comment by dxdesjardins** [comment]
On 1.0.1, when I export to inch, it creates a STEP with units of mm.  When I export to mm, it creates a STEP with units of inch.

**Comment by vernzimm** [comment]
> On 1.0.1, when I export to inch, it creates a STEP with units of mm. When I export to mm, it creates a STEP with units of inch.

Not sure if something was changed since then. Still using the same version of OCCT, but with weekly build 42180 it reliably exports in mm, regardless of what unit you are using to model or have set to export. Just make sure step export is set to mm and model will be the right size.

**Comment by FEA-eng** [comment]
Still relevant

```
OS: Windows 11 build 26200
Architecture: x86_64
Version: 1.2.0dev.20251223 (Git shallow)
Build date: 2025/12/23 18:19:32
Build type: Release
Branch: (HEAD detached at 065fea4)
Hash: 065fea473f080691988bb11092759bf40f3c46fe
Python 3.11.14, Qt 6.8.3, Coin 4.0.3, Vtk 9.3.1, boost 1_86, Eigen3 3.4.0, PySide 6.8.3
shiboken 6.8.3, xerces-c 3.3.0, IfcOpenShell 0.8.2, OCC 7.8.1
Locale: Polish/Poland (pl_PL)
Navigation Style/Orbit Style/Rotation Mode: Blender/Rounded Arcball/Window center
Stylesheet/Theme/QtStyle: FreeCAD.qss/FreeCAD Light/
Logical DPI/Physical DPI/Pixel Ratio: 96/81.5973/1
```

**Comment by vernzimm** [comment]
To be fair... I have it set to mm export, which does work correctly, and it's not really clear why it should have a unit selection in step format anyway. Maybe just set it to mm default always and be done with it?

**Comment by cweickhmann** [comment]
Main reason are is subsequent tools, e.g. numerical tool that expect SI, i. e. meters, as base units. And the problem isn't that it's one unit or the other but the fact that when you explicitly select one, you expect to get that one.

28 déc. 2025 23:26:28 vernzimm ***@***.***>:

**Comment by vernzimm** [comment]
@cweickhmann yeah, that's what I'm saying... why did they even add units to the file format in the first place? Just make it mm by definition, end of story.

I saw this was added to that OCCT Liaison project. Are they currently working to pull in new version of OCCT?

**Comment by ipatch** [comment]
unable able to reproduce with,

```
OS: Fedora Linux Asahi Remix 40 (Forty) (tty)
Architecture: arm64
Version: 1.2.0dev.44840 (Git)
Build date: 2025/12/30 17:40:39
Build type: Release
Branch: main
Hash: f5759e580b6330906beca12b4be419b030a088ab
Python 3.13.10, Qt 6.9.3, Coin 4.0.6, Vtk 9.5.2, boost 1_89, Eigen3 5.0.1, PySide 6.9.3
shiboken 6.9.3, SMESH 7.7.1.0, xerces-c 3.3.0, OCC 7.9.2
Locale: English/United States (en_US)
Navigation Style/Orbit Style/Rotation Mode: Blender/Rounded Arcball/Drag at cursor
Stylesheet/Theme/QtStyle: FreeCAD.qss/FreeCAD Light/
Logical DPI/Physical DPI/Pixel Ratio: 96/95.958/1
```

see the below comments,

https://github.com/FreeCAD/FreeCAD/issues/20546#issuecomment-3702885786

**Comment by vernzimm** [comment]
@ipatch I tracked the issue down to a known bug in OCCT 7.8.1 [OCCT Issue 354](https://github.com/Open-Cascade-SAS/OCCT/issues/354) that they had planned to fix with 7.9 that was scheduled for release in Feb 25, but we've just been rolling with 7.8.1 haha. I'm not sure why. Are you able to just build with 7.9.2? I was under the impression it would take a lot of work to bring it in, but if not...

**Comment by ipatch** [comment]
i build and install freecad locally using cmake + ninja from a CLI, and i pull-in / use the required deps built via homebrew. i believe right i only have occ v7.9.2 installed on my asahi linux box as that is the upstream homebrew-core version of occ. in theory i could install another version of occ to test with, but that would require some additional work.

**Comment by cweickhmann** [comment]
Hooray, that's good news. I'll check if I can build it like this too. Happy new year everyone, btw.

1 janv. 2026 00:20:56 Chris ***@***.***>:

**Comment by vernzimm** [comment]
@ipatch yeah, don't worry about testing with other versions, it's definitely been fixed. I'm more curious what was required to build with 7.9.2, and how can we get main to use it?

@cweickhmann thanks :)

**Comment by vernzimm** [comment]
@luzpaz see my earlier [reply](https://github.com/FreeCAD/FreeCAD/issues/18078#issuecomment-3702996832). I believe it was fixed in 7.9.0. I have the link there for the issue on OCCT repo.

**Comment by luzpaz** [comment]
If this is fixed in OCC 7.9.0 then this is a packaging related ticket.

**Comment by stefano2734** [comment]
Can be marked with label to 7.9.0 and higher.
Here important for users of occt 7.8.1 which is normal in 1.1.x and normal 26.3 dev versions wit occt 7.8.1.

**Comment by luzpaz** [comment]
This is fixed upstream. Users will see the fix when #32431 is implemented
